Speculation Rules: fix referrer-policy tests

yoavweiss opened the pull request:
#### 38bdd513e152d3f58ddea624652f00cc1e314b97
<pre>
Speculation Rules: fix referrer-policy tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=300608">https://bugs.webkit.org/show_bug.cgi?id=300608</a>

Reviewed by Alex Christensen.

The current referrer-policy speculation rules tests are failing for multiple reasons:
* get-host-info is out-of-date and they use undefined parts of it.
* Invalid referrer-policy doesn&apos;t currently prevent a prefetch from happening.

This PR fixes that by manually importing the missing parts of get-host-info, and by aligning the implementation to the spec.

No new tests, but fixing expectations on existing tests.

* LayoutTests/TestExpectations: Remove referrer-policy as expected failure.
* LayoutTests/imported/w3c/web-platform-tests/common/get-host-info.sub.js:
(get_host_info): Update to the latest.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/cross-origin-cookies.https-expected.txt: Adjusted expectations.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/out-of-document-rule-set.https.html: Removed no-longer needed difference from upstream WPTs.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/referrer-policy-from-rules.https_3-3-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/referrer-policy-from-rules.https_4-4-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/referrer-policy-from-rules.https_6-6-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/referrer-policy-from-rules.https_7-7-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/referrer-policy-not-accepted.https_2-2-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/referrer-policy-not-accepted.https_3-last-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/user-pass.https_cross-origin=true-expected.txt: Adjusted expectations.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/speculation-tags/cross-site-prefetch.https-expected.txt: Adjusted expectations.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/speculation-tags/cross-site-to-same-site-redirection-prefetch.https-expected.txt: Adjusted expectations.
* LayoutTests/platform/gtk/TestExpectations: Remove referrer-policy as expected failure.
* LayoutTests/platform/win/TestExpectations: Remove referrer-policy as expected failure.
* LayoutTests/platform/wpe/TestExpectations: Remove referrer-policy as expected failure.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::prefetch): Pass ReferrerPolicy instead of String.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/SpeculationRulesMatcher.h:
* Source/WebCore/html/HTMLAnchorElement.cpp:
(WebCore::HTMLAnchorElement::setShouldBePrefetched): Pass ReferrerPolicy instead of String.
(WebCore::HTMLAnchorElement::checkForSpeculationRules): Pass ReferrerPolicy instead of String.
* Source/WebCore/html/HTMLAnchorElement.h:
* Source/WebCore/loader/DocumentPrefetcher.cpp: Pass ReferrerPolicy instead of String.
(WebCore::makePrefetchRequest):
(WebCore::DocumentPrefetcher::prefetch): Pass ReferrerPolicy instead of String.
* Source/WebCore/loader/DocumentPrefetcher.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::prefetch): Pass ReferrerPolicy instead of String.
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/SpeculationRules.cpp:
(WebCore::parseSingleRule): Align implementation to the spec.
* Source/WebCore/loader/SpeculationRules.h:

Canonical link: <a href="https://commits.webkit.org/301476@main">https://commits.webkit.org/301476@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/329c340ec9473bca0c0e4c831719aab637cd853d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126083 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45736 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36521 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132895 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77886 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127954 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46407 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54281 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96015 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64137 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129031 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37126 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112763 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76504 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36029 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30945 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76367 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106908 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31167 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135602 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52839 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40572 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104520 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53297 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108980 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104236 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26562 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49632 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27959 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50222 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52736 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58570 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52064 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55410 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53774 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/loader/SpeculationRules.h:
```diff
@@ -25,6 +25,7 @@
 
 #pragma once
 
+#include "ReferrerPolicy.h"
```
This can be a forward declaration instead an include from a header that slows down the build.

yoavweiss replied:
Done

achristensen07 commented on Source/WebCore/dom/SpeculationRulesMatcher.h:
```diff
@@ -25,6 +25,7 @@
 
 #pragma once
 
+#include "ReferrerPolicy.h"
```
Forward declaration.

yoavweiss replied:
Done! For some reason I thought that members need to have their full declaration..

yoavweiss committed:
Speculation Rules: fix referrer-policy tests
https://bugs.webkit.org/show_bug.cgi?id=300608

Reviewed by Alex Christensen.

The current referrer-policy speculation rules tests are failing for multiple reasons:
* get-host-info is out-of-date and they use undefined parts of it.
* Invalid referrer-policy doesn't currently prevent a prefetch from happening.

This PR fixes that by manually importing the missing parts of get-host-info, and by aligning the implementation to the spec.

No new tests, but fixing expectations on existing tests.

* LayoutTests/TestExpectations: Remove referrer-policy as expected failure.
* LayoutTests/imported/w3c/web-platform-tests/common/get-host-info.sub.js:
(get_host_info): Update to the latest.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/cross-origin-cookies.https-expected.txt: Adjusted expectations.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/out-of-document-rule-set.https.html: Removed no-longer needed difference from upstream WPTs.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/referrer-policy-from-rules.https_3-3-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/referrer-policy-from-rules.https_4-4-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/referrer-policy-from-rules.https_6-6-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/referrer-policy-from-rules.https_7-7-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/referrer-policy-not-accepted.https_2-2-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/referrer-policy-not-accepted.https_3-last-expected.txt: Progression.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/prefetch/user-pass.https_cross-origin=true-expected.txt: Adjusted expectations.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/speculation-tags/cross-site-prefetch.https-expected.txt: Adjusted expectations.
* LayoutTests/imported/w3c/web-platform-tests/speculation-rules/speculation-tags/cross-site-to-same-site-redirection-prefetch.https-expected.txt: Adjusted expectations.
* LayoutTests/platform/gtk/TestExpectations: Remove referrer-policy as expected failure.
* LayoutTests/platform/win/TestExpectations: Remove referrer-policy as expected failure.
* LayoutTests/platform/wpe/TestExpectations: Remove referrer-policy as expected failure.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::prefetch): Pass ReferrerPolicy instead of String.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/SpeculationRulesMatcher.h:
* Source/WebCore/html/HTMLAnchorElement.cpp:
(WebCore::HTMLAnchorElement::setShouldBePrefetched): Pass ReferrerPolicy instead of String.
(WebCore::HTMLAnchorElement::checkForSpeculationRules): Pass ReferrerPolicy instead of String.
* Source/WebCore/html/HTMLAnchorElement.h:
* Source/WebCore/loader/DocumentPrefetcher.cpp: Pass ReferrerPolicy instead of String.
(WebCore::makePrefetchRequest):
(WebCore::DocumentPrefetcher::prefetch): Pass ReferrerPolicy instead of String.
* Source/WebCore/loader/DocumentPrefetcher.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::prefetch): Pass ReferrerPolicy instead of String.
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/SpeculationRules.cpp:
(WebCore::parseSingleRule): Align implementation to the spec.
* Source/WebCore/loader/SpeculationRules.h:

Canonical link: https://commits.webkit.org/301476@main
